### DISCUSSION TITLE
Expected behavior when comparing UTF8STRING and PRINTABLESTRING during DN matching?

### DISCUSSION THREAD

**Discussion by levicki** [openssl/openssl#24593] [Q&A]
Consider this scenario:

![image](https://user-images.githubusercontent.com/16415478/169032053-23166637-2f54-4af4-b3b6-346d9d3fa6e0.png)

The two strings are obviously identical so the error message is somewhat misleading.

So where is the difference then?

This is the CA certificate:
![image](https://user-images.githubusercontent.com/16415478/169034240-0a9f3998-db52-4139-aa33-825321fdcf0a.png)

And this is the CSR:
![image](https://user-images.githubusercontent.com/16415478/169032618-24da7f23-dd21-4817-acea-92a029f1d560.png)

I think some clarification is needed on this.

1. Is this field comparison expected to fail even when `PRINTABLESTRING` and `UTF8STRING` values are byte-identical strings (i.e. when both consist of 7-bit ASCII characters where `memcmp()` result would be `0`)?
2. If it is supposed to fail, why the error message does not reflect that by saying that field **ASN.1 encodings** do not match instead of saying that the fields don't match?

I am aware that I can just ignore this mismatch by setting `stateOrProvinceName` to be optional in the OpenSSL config policy section, but I would like to know whether this is an error or expected behavior. If it is expected I think it should have a better error message.

**EDIT:**

In my opinion, if the strings are byte-identical, the ASN.1 encoding mismatch should be a warning, not a failure.

**Rationale:**

The encoding can differ due to all sorts of different 3rd party CSR generation tools (such as in this case [APC Network Management Card Security Wizard CLI Utility](https://www.apc.com/shop/us/en/products/APC-Network-Management-Card-Security-Wizard-CLI-Utility-v1-0-1/P-SFNMCSECCLI101) or built-in CSR generation implementations on various network appliances where end user has no control over CSR field encoding.

So, instead of having to either make the field optional in the CA policy and thus allow for actual value mismatch or instead of having to reissue a full CA certificate chain used for signing once you hit an incompatible CSR generation tool / network appliance, I believe it would be better to try to match using byte compare and if the values match just issue a warning for mismatched ASN.1 encoding and ask the user whether to allow that while still refusing to sign the CSR if the values differ.

Note that I am by no means a security expert so I cannot assess the potential security implications of such a change and the opinion stated above is given solely from an OpenSSL end-user viewpoint. That is also the reason why I created this issue as a question, rather than a feature request or a bug report.

**Comment by kaduk** [comment]
Leaving some notes here since I found a couple potentially relevant references, even though I did not reach a definitive conclusion in the time allotted. https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.6 is quite explicit that there is a requirement for equality of encoding type as well as contents in specific situations that require equality (e.g., when a CA issues certificates, the Subject of the issuing certificate must exactly match the Issuer in the issued certificates).  However, later parts of RFC 5280 refer to RFC 4518 for string preparation, and the preparation step (https://datatracker.ietf.org/doc/html/rfc4518#section-2.1) does include a "transcode the string value to Unicode" step, implying that some variation might be acceptable in other situations.

As to why the current behavior and error message is the way it is, I submit that the logic is essentially unchanged from the original SSLeay code, which did not attempt to consider such subtleties.

**Comment by levicki** [comment]
@kaduk @t8m 
> ... does include a "transcode the string value to Unicode" step, implying that some variation might be acceptable in other situations.

Thanks for taking the time to investigate.

The above quote makes sense to me provided that the proper validation is performed on `PRINTABLESTRING` value before conversion to UTF-8 or Unicode to make sure it does not contain byte values with 8th bit set which could turn into different characters after transcoding.

Perhaps there is some room for improvement of the current handling once you understand where such conversions are allowed by the standard?

**Comment by Slydder** [comment]
Hello all,

Was just wondering what, if any, is the plan for fixing this issue. And it is obviously an issue that does need to be addressed. Currently if requests from downstream are generated using -subj will somehow get you a mix of UTF8 and printablestrings (at least here I cat the countryName as PRINTABLESTRING and the rest as UTF8) and thus are failing automated issueing then this is an issue seeing as each time an admin is required to manually adjust to issue the cert.

Oh, and this is still an issue as of 3.0.2

thx,
chuck

**Comment by vdukhovni** [comment]
1. The sensible thing for a modern CA to do is to simply effectively **omit** subject DNs from issued EE certificates by setting the subject DN to the empty sequence (e.g. OpenSSL `-subj /` option). This is already the practice, e.g., for DNS-ID certificates, when the hostname in question is longer than the 64 byte limit on `CommonName`. Then questions of DN component encoding do not arise, and the CA policy can ignore the entire DN in the CSR, and check only the SAN(s).
2. The name binding of subject to key in the certificate is then just for each of the SANs. These are required to be encoded per the requirements for the particular SAN type (e.g. IA5String for DNS-ID).
3. The OpenSSL `ca(1)` command-line utility implements a *very* bare-bones (minimal) CA, that may be sufficient for intramural use in a small organisation, but is not expected to be used by CAs operated to serve commercial clients. If `ca(1)` is somewhat unforgiving in heterogenous environments, that should have no impact on commercial CAs. Multiple workarounds are available to operators of intramural CAs in end-organisations.

As for identifying an all-ASCII PrintableString with a corresponding all-ASCII UTF8String, when performing DN comparisons, I'm not sure this would be a good idea. Even if this greases the wheels of certificate issuance, it is unclear which encoding of that component should be in the issued certificate, the requested encoded or the CAs encoding? Either way, problems may later show up in evaluation of DirectoryName name constraints in processing of revocation requests, ...
